### PR TITLE
solution to hw3-2

### PR DESCRIPTION
**Pull request by syevtushok**

### CODE REVIEW

File: src/main/java/core/basesyntax/Rectangle.java [hunk 1]
```diff
@@ -27,4 +27,16 @@ public Integer getLength() {
     public String getColor() {
         return color;
     }
+
+    @Override
+    public int hashCode() {
+        return 27 * getLength() * getWidth();
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Rectangle.java]: how about color?
we shouldn't depend on value of fields when we find hash code

File: src/main/java/core/basesyntax/Rectangle.java [hunk 2]
```diff
@@ -27,4 +27,16 @@ public Integer getLength() {
...
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return this.getColor().equals(getColor())
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Rectangle.java]: how about other checking? We should have 4 checking here

File: src/main/java/core/basesyntax/Rectangle.java [hunk 3]
```diff
@@ -27,4 +27,31 @@ public Integer getLength() {
     public String getColor() {
         return color;
     }
+
+    @Override
+    public int hashCode() {
+        return 31 * getLength() * getWidth()
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Rectangle.java]: what will happen if width or length ==0?
try do smth like this
https://www.baeldung.com/java-hashcode

File: src/main/java/core/basesyntax/Rectangle.java [hunk 4]
```diff
@@ -27,4 +27,31 @@ public Integer getLength() {
...
+            * (getColor() != null ? getColor().length() : 2);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+
+        if (obj instanceof Rectangle) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Rectangle.java]: what is the difference between method `instanceof` and `getClass()`?

File: src/main/java/core/basesyntax/Rectangle.java [hunk 5]
```diff
@@ -27,4 +27,31 @@ public Integer getLength() {
...
+            if (!this.getColor().equals(rectangle.getColor())) {
+                return false;
+            }
+            if (!this.getWidth().equals(rectangle.getWidth())) {
+                return false;
+            }
+            if (!this.getLength().equals(rectangle.getLength())) {
+                return false;
+            }
+        }
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Rectangle.java]: Can we write it with && ?

File: src/main/java/core/basesyntax/Rectangle.java (hunk 4 above)
Comment (syevtushok) [reply src/main/java/core/basesyntax/Rectangle.java]: Oh, I understood it. I'll change it.

File: src/main/java/core/basesyntax/Rectangle.java (hunk 3 above)
Comment (Sofasmile) [reply src/main/java/core/basesyntax/Rectangle.java]: what will happen if length or width is 0?

can we use a hashcode of width, length and color and other algorithm of calculation?

File: src/main/java/core/basesyntax/Rectangle.java [hunk 6]
```diff
@@ -27,4 +27,22 @@ public Integer getLength() {
...
+    public int hashCode() {
+        return 31 * getLength() * getWidth()
+            * (getColor() != null ? getColor().length() : 2);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Rectangle.java:42]: how about checking on null?

File: src/main/java/core/basesyntax/Rectangle.java [hunk 7]
```diff
@@ -27,4 +27,22 @@ public Integer getLength() {
...
+        Rectangle rectangle = (Rectangle) obj;
+        return ((Rectangle) obj).getClass().equals(rectangle.getClass())
+            && this.getColor().equals(rectangle.getColor())
+            && this.getWidth().equals(rectangle.getWidth())
+            && this.getLength().equals(rectangle.getLength());
+    }
+}
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Rectangle.java]: code style !

File: src/main/java/core/basesyntax/Rectangle.java [hunk 8]
```diff
@@ -27,4 +27,25 @@ public Integer getLength() {
     public String getColor() {
         return color;
     }
+
+    @Override
+    public int hashCode() {
+        return 31 * color.hashCode() * width.hashCode() * length.hashCode();
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Rectangle.java]: what will happen if length or width is 0 or color = "" ?

**Review by Sofasmile — COMMENTED** [review]
Good work!
Let's fix the latest comment :)

**Review by Sofasmile — APPROVED** [review]

**Pull request closed without merging**